[WIP] Ceph RBD dynamic provisioning

rootfs opened the pull request:
This replaces #19710.
- Create storage class and PVs works. Provisioner parameters have Ceph client credential.
- Cannot delete PVC now. Because RBD requires Ceph client credential. I haven't figured out how to get the provisioner parameter in deleter yet (or is it possible? @jsafrane )

@kubernetes/sig-storage

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30959)

<!-- Reviewable:end -->

Commit:
initial drop

Signed-off-by: Huamin Chen <hchen@redhat.com>

Commit:
support storage class in Ceph RBD volume

Signed-off-by: Huamin Chen <hchen@redhat.com>

johscheuer commented:
Hi @rootfs nice to see your work! Just some questions from my side (if I'm wrong at any point please correct me :))

1.) This PR requires Ceph installed on all client nodes and on the master nodes (https://github.com/kubernetes/kubernetes/pull/30959/files#diff-36f18e327f36d95eb1333c4a18781184R327) ? 

1.1.) If Ceph has a REST API (and the PR would use it) there would be no need for installed Ceph tools on the Master node(s) ?

2.) I assume that the Controller (Kube-Master) triggers the creation of the PV/Image? So would it be possible to move the admin credentials (username + password) into a secret in a namespace e.g. `kube-system`? (This would assume that all PV are created with the same admin account)

Thanks!

jsafrane committed:
Add StorageClass.Parameters to volume plugin Deleter interface.

StorageClass may contain information necessary to delete a volume, e.g.
address of remote Ceph server and thus must be passed not only to
provisioners, but also to deleters.

jsafrane commented:
> Cannot delete PVC now. Because RBD requires Ceph client credential. I haven't figured out how to get the provisioner parameter in deleter yet (or is it possible? @jsafrane )


Soon, see #30986

rootfs commented:
@johscheuer 
Yes, `ceph-common` package (with `rbd` binary and probably `rados` as well if we decide to dynamically create Ceph client and assign privileges later) on kube master so provisioner can run the command. `rbd` must be installed on kube node to map rbd images.

This PR assumes no Ceph REST available. It uses `rbd create` and `rbd rm` to provision and delete images.

Storing admin credential as a secret in a well known namespace is certain an option. There is generic discussion at #30897

Commit:
pull from PR #30986

rootfs commented:
Pulled #30986 and now PV can be deleted.

Commit:
use secret instead of plaintext for Ceph key

Signed-off-by: Huamin Chen <hchen@redhat.com>

rootfs commented:
In the storage class, Ceph key is no longer in plaintext. A sample storage class is
```yaml
apiVersion: extensions/v1beta1
kind: StorageClass
metadata:
   name: slow
provisioner: kubernetes.io/rbd
parameters:
    monitors: 10.16.153.105:6789
    adminID: admin
    adminSecretName: ceph-secret-admin
    adminSecretNamespace: kube-system
    pool: kube
    userId: kube
    secretName: ceph-secret-user
```

elsonrodriguez commented:
@rootfs Provisioning looks good! 

However I wasn't able to mount:

```
Aug 21 00:04:20 node-f5d9368b.harbor0.group.company.com kubelet[151358]: I0821 00:04:20.805749  151358 rbd_util.go:241] rbd: map error exit status 6 2016-08-21 00:04:10.802512 7fdbbe321100 -1 did not load config file, using default settings.
Aug 21 00:04:20 node-f5d9368b.harbor0.group.company.com kubelet[151358]: 2016-08-21 00:04:10.812965 7fdbbe321100 -1 auth: unable to find a keyring on /etc/ceph/ceph.client.admin.keyring,/etc/ceph/ceph.keyring,/etc/ceph/keyring,/etc/ceph/keyring.bin: (2) No such file or directory
Aug 21 00:04:20 node-f5d9368b.harbor0.group.company.com kubelet[151358]: rbd: sysfs write failed
Aug 21 00:04:20 node-f5d9368b.harbor0.group.company.com kubelet[151358]: rbd: map failed: (6) No such device or address
Aug 21 00:04:20 node-f5d9368b.harbor0.group.company.com kubelet[151358]: E0821 00:04:20.805809  151358 disk_manager.go:61] failed to attach disk
Aug 21 00:04:20 node-f5d9368b.harbor0.group.company.com kubelet[151358]: E0821 00:04:20.805821  151358 rbd.go:408] rbd: failed to setup mount /var/lib/kubelet/pods/d9a5aedf-6768-11e6-8a4f-001e677fb1d7/volumes/kubernetes.io~rbd/pvc-33cbaa5b-6768-11e6-8a4f-001e677fb1d7 rbd: map failed exit status 6 2016-08-21 00:04:10.802512 7fdbbe321100 -1 did not load config file, using default settings.
Aug 21 00:04:20 node-f5d9368b.harbor0.group.company.com kubelet[151358]: 2016-08-21 00:04:10.812965 7fdbbe321100 -1 auth: unable to find a keyring on /etc/ceph/ceph.client.admin.keyring,/etc/ceph/ceph.keyring,/etc/ceph/keyring,/etc/ceph/keyring.bin: (2) No such file or directory
Aug 21 00:04:20 node-f5d9368b.harbor0.group.company.com kubelet[151358]: rbd: sysfs write failed
Aug 21 00:04:20 node-f5d9368b.harbor0.group.company.com kubelet[151358]: rbd: map failed: (6) No such device or address
Aug 21 00:04:20 node-f5d9368b.harbor0.group.company.com kubelet[151358]: E0821 00:04:20.805905  151358 nestedpendingoperations.go:253] Operation for "\"kubernetes.io/rbd/d9a5aedf-6768-11e6-8a4f-001e677fb1d7-pvc-33cbaa5b-6768-11e6-8a4f-001e677fb1d7\" (\"d9a5aedf-6768-11e6-8a4f-001e677fb1d7\")" failed. No retries permitted until 2016-08-21 00:06:20.805864537 -0700 PDT (durationBeforeRetry 2m0s). Error: MountVolume.SetUp failed for volume "kubernetes.io/rbd/d9a5aedf-6768-11e6-8a4f-001e677fb1d7-pvc-33cbaa5b-6768-11e6-8a4f-001e677fb1d7" (spec.Name: "pvc-33cbaa5b-6768-11e6-8a4f-001e677fb1d7") pod "d9a5aedf-6768-11e6-8a4f-001e677fb1d7" (UID: "d9a5aedf-6768-11e6-8a4f-001e677fb1d7") with: rbd: map failed exit status 6 2016-08-21 00:04:10.802512 7fdbbe321100 -1 did not load config file, using default settings.
Aug 21 00:04:20 node-f5d9368b.harbor0.group.company.com kubelet[151358]: 2016-08-21 00:04:10.812965 7fdbbe321100 -1 auth: unable to find a keyring on /etc/ceph/ceph.client.admin.keyring,/etc/ceph/ceph.keyring,/etc/ceph/keyring,/etc/ceph/keyring.bin: (2) No such file or directory
Aug 21 00:04:20 node-f5d9368b.harbor0.group.company.com kubelet[151358]: rbd: sysfs write failed
Aug 21 00:04:20 node-f5d9368b.harbor0.group.company.com kubelet[151358]: rbd: map failed: (6) No such device or address
```
I tried putting the keyring in /etc/ceph, and reduced the verbosity of the error:

```
Aug 21 00:08:23 node-f5d9368b.harbor0.group.company.com kubelet[151358]: I0821 00:08:23.190922  151358 rbd_util.go:230] rbd: map mon ceph-mon.ceph.svc.harbor0.group.company.com:6789
Aug 21 00:08:23 node-f5d9368b.harbor0.group.company.com kubelet[151358]: I0821 00:08:23.241221  151358 rbd_util.go:241] rbd: map error exit status 6 2016-08-21 00:08:23.206682 7f04da321100 -1 did not load config file, using default settings.
Aug 21 00:08:23 node-f5d9368b.harbor0.group.company.com kubelet[151358]: rbd: sysfs write failed
Aug 21 00:08:23 node-f5d9368b.harbor0.group.company.com kubelet[151358]: rbd: map failed: (6) No such device or address
Aug 21 00:08:23 node-f5d9368b.harbor0.group.company.com kubelet[151358]: E0821 00:08:23.241289  151358 disk_manager.go:61] failed to attach disk
Aug 21 00:08:23 node-f5d9368b.harbor0.group.company.com kubelet[151358]: E0821 00:08:23.241301  151358 rbd.go:408] rbd: failed to setup mount /var/lib/kubelet/pods/d9a5aedf-6768-11e6-8a4f-001e677fb1d7/volumes/kubernetes.io~rbd/pvc-33cbaa5b-6768-11e6-8a4f-001e677fb1d7 rbd: map failed exit status 6 2016-08-21 00:08:23.206682 7f04da321100 -1 did not load config file, using default settings.
Aug 21 00:08:23 node-f5d9368b.harbor0.group.company.com kubelet[151358]: rbd: sysfs write failed
Aug 21 00:08:23 node-f5d9368b.harbor0.group.company.com kubelet[151358]: rbd: map failed: (6) No such device or address
Aug 21 00:08:23 node-f5d9368b.harbor0.group.company.com kubelet[151358]: E0821 00:08:23.241425  151358 nestedpendingoperations.go:253] Operation for "\"kubernetes.io/rbd/d9a5aedf-6768-11e6-8a4f-001e677fb1d7-pvc-33cbaa5b-6768-11e6-8a4f-001e677fb1d7\" (\"d9a5aedf-6768-11e6-8a4f-001e677fb1d7\")" failed. No retries permitted until 2016-08-21 00:10:23.241375676 -0700 PDT (durationBeforeRetry 2m0s). Error: MountVolume.SetUp failed for volume "kubernetes.io/rbd/d9a5aedf-6768-11e6-8a4f-001e677fb1d7-pvc-33cbaa5b-6768-11e6-8a4f-001e677fb1d7" (spec.Name: "pvc-33cbaa5b-6768-11e6-8a4f-001e677fb1d7") pod "d9a5aedf-6768-11e6-8a4f-001e677fb1d7" (UID: "d9a5aedf-6768-11e6-8a4f-001e677fb1d7") with: rbd: map failed exit status 6 2016-08-21 00:08:23.206682 7f04da321100 -1 did not load config file, using default settings.
Aug 21 00:08:23 node-f5d9368b.harbor0.group.company.com kubelet[151358]: rbd: sysfs write failed
Aug 21 00:08:23 node-f5d9368b.harbor0.group.company.com kubelet[151358]: rbd: map failed: (6) No such device or address
```

I thought it might be a DNS issue but things look ok:

```
nslookup  ceph-mon.ceph.svc.harbor0.group.company.com
Server:		10.54.9.219
Address:	10.54.9.219#53

Non-authoritative answer:
Name:	ceph-mon.ceph.svc.harbor0.group.company.com
Address: 10.9.220.4
Name:	ceph-mon.ceph.svc.harbor0.group.company.com
Address: 10.9.219.2
Name:	ceph-mon.ceph.svc.harbor0.group.company.com
Address: 10.9.216.6
```
Any clues?

Also would it be possible to take the keyring from a secret as well?

rootfs commented:
@elsonrodriguez thank you for testing!

Can you share your rbd storage class? It seems you were using keyring instead of secret. But here keyring is not supported.

elsonrodriguez commented:
I only stuck a keyring manually on the host because I saw it in the error message and thought it might help (it didn't).

I'm using the same secret for both admin and user. I know the secret works because it provisioned the RBD volume just fine.

Here is my storage class:
```
apiVersion: extensions/v1beta1
kind: StorageClass
metadata:
   name: ceph-default
provisioner: kubernetes.io/rbd
parameters:
    monitors: ceph-mon.ceph.svc.harbor0.group.company.com:6789
    adminID: admin
    adminSecretName: ceph-client-key
    adminSecretNamespace: ceph
    pool: rbd
    userId: admin
    secretName: ceph-client-key
```

rootfs commented:
@elsonrodriguez yes, that looks right to me. Is `ceph-client-key`  also in the namespace where you create your pod?

rootfs commented:
@leseb

elsonrodriguez commented:
@rootfs It mounts!

It was an issue on my end. I had to disable `exclusive-lock` on the volume. There was no lock on the volume but my system refused to mount it unless I disabled the feature.

rootfs commented:
@elsonrodriguez cool!

Next step I am looking for input on what rbd image options we should support (strip unit, order, layer, share, etc). Would love to hear use cases.

Commit:
temporary patch to gluster deleter

Signed-off-by: Huamin Chen <hchen@redhat.com>

elsonrodriguez commented on examples/experimental/persistent-volume-provisioning/rbd/pod.yaml:
```diff
@@ -0,0 +1,23 @@
+    spec:
+      containers:
+      - name: server
+        image: ngnix
```
Typo here, needs to be nginx instead of ngnix

Commit:
fix image name

Signed-off-by: Huamin Chen <hchen@redhat.com>

elsonrodriguez commented:
@rootfs I'm Having trouble reproducing success with the provisioner.

Occasionally I will get a claim that won't create a PV or RBD.

Sometimes a claim will create an RBD but no PV to match.

Partial Success:
```
$ kubectl  get pvc
NAME      STATUS    VOLUME                                     CAPACITY   ACCESSMODES   AGE
claim1    Pending                                                                       37m
claim2    Bound     pvc-3519eae6-689b-11e6-8a4f-001e677fb1d7   10Gi       RWO           1h
claim3    Bound     pvc-92456f61-68a5-11e6-8a4f-001e677fb1d7   50Gi       RWO           36m
```

```
$ rbd ls
kubernetes-dynamic-pvc-031e00eb-68a5-11e6-ab45-001e677fb1d7
kubernetes-dynamic-pvc-351a763a-689b-11e6-ab45-001e677fb1d7
kubernetes-dynamic-pvc-9246759c-68a5-11e6-ab45-001e677fb1d7
```

No success:

```
$ kubectl  get pvc
NAME      STATUS    VOLUME    CAPACITY   ACCESSMODES   AGE
claim1    Pending                                      3m
claim2    Pending                                      1m
claim3    Pending                                      53s
```
```
rbd ls
```

Trying to debug, but if you have any hints they'd be appreciated.

elsonrodriguez commented:
@rootfs There was a networking issue with my cluster, it was causing `rbd` to hang sometimes. Things now provision without issue, and mounts when the right features are disabled.

It looks good so far and the issues were user error related. Thanks for the amazing work!

+1 on your earlier idea to let the storage class define the rbd features, that sounds like a win.

Commit:
Merge branch 'rbd-prov3' into rbd-prov2

rootfs commented:
closing now, see final PR at #31251
